src/wamr/wamr.cc

```diff
@@ -0,0 +1,631 @@
+// Copyright 2016-2019 Envoy Project Authors
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "include/proxy-wasm/wamr.h"
+#include "include/proxy-wasm/wasm_vm.h"
+
+#include <array>
+#include <cassert>
+#include <cstring>
+#include <iomanip>
+#include <memory>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "src/wamr/types.h"
+#include "wasm_c_api.h"
+#include "src/common/bytecode_util.h"
+
+namespace proxy_wasm {
+namespace wamr {
+
+struct HostFuncData {
+  HostFuncData(std::string name) : name_(std::move(name)) {}
+
+  std::string name_;
+  WasmFuncPtr callback_;
+  void *raw_func_;
+  WasmVm *vm_;
+};
+
+using HostFuncDataPtr = std::unique_ptr<HostFuncData>;
+
+wasm_engine_t *engine() {
+  static const auto engine = WasmEnginePtr(wasm_engine_new());
+  return engine.get();
+}
+
+class Wamr : public WasmVm {
+public:
+  Wamr() {}
+
+  std::string_view runtime() override { return "envoy.wasm.runtime.wamr"; }
```

Comment:
This should return `"wamr"`, since Proxy-Wasm C++ Host is (in theory) proxy-agnostic.

The issue you've hit should be fixed in https://github.com/envoyproxy/envoy/pull/16263.